File: 02654. Minimum Number of Operations to Make All Array Elements Equal to 1/solution.go
/**
 * Problem: 2654. Minimum Number of Operations to Make All Array Elements Equal to 1
 *
 * Difficulty: Medium
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func gcd(a, b int) int {
	// Loop until the second number becomes zero
	for b != 0 {
		// Update 'a' to 'b' and 'b' to the remainder of 'a' divided by 'b'
		a, b = b, a%b
	}

	// If the result 'a' is negative, return its absolute value
	if a < 0 {
		return -a
	}

	// Return the greatest common divisor
	return a
}

func minOperations(nums []int) int {
	// Get the length of the input array
	n := len(nums)

	// Initialize a counter for the number of ones in the array
	ones := 0

	// Count the number of ones in the array
	for _, x := range nums {
		if x == 1 {
			ones++
		}
	}

	// If there are any ones, return the number of non-one elements
	if ones > 0 {
		return n - ones
	}

	// Initialize the gcd accumulator
	g := 0

	// Compute the gcd of all elements in the array
	for _, x := range nums {
		g = gcd(g, x)
	}

	// If the gcd of all elements is greater than 1, return -1
	if g > 1 {
		return -1
	}

	// Define a large constant to represent infinity
	const INF = int(1e9)

	// Initialize the best length to infinity
	best := INF

	// Iterate over all possible starting indices
	for i := 0; i < n; i++ {
		// Initialize the current gcd for the subarray
		cur := 0

		// Iterate over all possible ending indices
		for j := i; j < n; j++ {
			// Update the gcd for the current subarray
			cur = gcd(cur, nums[j])

			// If the gcd becomes 1, update the best length
			if cur == 1 {
				if length := j - i + 1; length < best {
					best = length
				}

				// Break out of the loop as we found a valid subarray
				break
			}
		}
	}

	// Return the total operations needed to make all elements 1
	return (best - 1) + (n - 1)
}

File: 02654. Minimum Number of Operations to Make All Array Elements Equal to 1/solution.cpp
/**
 * Problem: 2654. Minimum Number of Operations to Make All Array Elements Equal to 1
 *
 * Difficulty: Medium
 *
 * Language: C++
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

class Solution
{
public:
  int minOperations(vector<int> &nums)
  {
    // Get the size of the input array
    const int n = nums.size();

    // Initialize the result to the maximum integer value
    // and a counter for the number of elements equal to 1
    int res = INT_MAX, cnt1 = 0;

    // Count the number of elements in the array that are equal to 1
    for (int i = 0; i < n; ++i)
      cnt1 += (nums[i] == 1);

    // If there are any 1s in the array, return the number of
    // operations needed to make all elements equal to 1
    if (cnt1)
      return n - cnt1;

    // Iterate through the array to find the minimum number of
    // operations required to make all elements equal to 1
    for (int i = 0; i < n; ++i)
    {
      // Initialize the gcd (greatest common divisor) with the current element
      int g = nums[i];

      // Iterate through the rest of the array to calculate gcd
      for (int j = i + 1; j < n; ++j)
      {
        // Update the gcd with the current element
        g = __gcd(g, nums[j]);

        // If gcd becomes 1, calculate the minimum operations
        if (g == 1)
        {
          res = min(res, j - i + n - 1);
          break; // Break out of the inner loop as gcd is already 1
        }
      }
    }

    // If no valid operations were found, return -1; otherwise, return the result
    return res == INT_MAX ? -1 : res;
  }
};